### PR TITLE
add support for Step Predictor and Forecaster abstract Interfaces.

### PR DESCRIPTION
**Pull request by kartikangiras**
## Describe your changes

 Summary of the changes.
Introduce two new abstract base classes that form the foundation of the refactored Neural-LAM architecture:

- StepPredictor: Maps (X_{t-1}, X_t, F_t) → X_{t+1} to a single-step neural network abstraction (GNN, CNN)

- Forecaster: Maps (init_states, forcing, true_states) to forecast over a full time window  which is the forecasting strategy layer. 

Please also include relevant motivation and context. 

The current ARModelclass is a monolithic class that handles AR unrolling, batch management, loss computation, logging, and GNN forward passes all mixed together. This code change begins splitting those responsibilities into well-scoped, composable layers.

This enables:

- Clear separation of concerns (neural net architecture vs. forecasting strategy vs. Lightning orchestration).
- Support for future architectures (CNNs, ensemble models, non-AR strategies).
- Reusability of StepPredictor outside the training loop (inference, scripting).

List any dependencies that are required for this change. 
None
## Issue Link
relates to #49 (this is the first part of the entire implementation).
< Link to the relevant issue or task, if applicable > (e.g. `closes #00` or `solves #00`)

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by joeloskarsson** [comment]
Following my comment in #49 , let's focus efforts in refactoring the model hierarchy on #208. Please share any ideas and contribute there :) To avoid duplicates I will close this for now.
